### DISCUSSION TITLE
Duplicated input and unable to delete the first letter on GTK4 applications

### DISCUSSION THREAD

**Discussion by pineapplehunter** [google/mozc#931] [Q&A] [answered]
## Description
I am having issues when using mozc on gtk4 applications. The issues are,

1. When I type in something in Japanese and press enter, it creates a duplicate input after the input and cannot be deleted until I click somewhere else with the mouse.
2. When I type in something in Japanese, pressing backspace leaves one character and cannot be deleted until I click somewhere else with the mouse.

I am trying to figure out what is causing this behavior, but I don't have much knowledge of this program.
If there are any pointers or ways to debug the process, I am very happy to help. 

## Steps to reproduce
1. Open a gtk4 application
2. type something in Japanese
3. hit enter, or try to delete

## Expected behavior
- When enter is hit, it should not be duplicated
- When deleting, it should delete everything

## Actual behavior
See Screenshots.

## Screenshots
 
[The enter case](https://github.com/google/mozc/assets/8869894/ec54d2de-4295-4250-a81d-98b8eff853fe)
[The delete case](https://github.com/google/mozc/assets/8869894/ca4a6495-1b8d-44eb-9cc3-4d3de5af63af)

## Version or commit-id
Tested both on mozc built on tag 2.29.5374.102 and latest git.

## Environment
 - OS: NixOS unstable
 - IMF (for Linux) : IBus
 - Related Applications: All GTK4 Applications. GTK3 applications and other applications with different frameworks are not affected.

## Investigations
 - Whether this issue happens on both Ibus and other IMF (e.g. Fcitx, uim).
   + no, there were no issues on fcitx5 (different mozc version)
 - Whether this issue happens on other IMEs (e.g. Anthy, SKK).
   + no, ibus with anthy seem to be working correctly
 - What applications this issue happens on (e.g. Chromium, gedit).
   + GTK4 or libadwaita applications?
 - What applications this issue does not happen on (e.g. Chromium, gedit).
   + Firefox, chrome, gtk3 applications, qt applications
 - (optional) What versions or commit-ids this issue did not happen on
   (e.g. Mozc-2.28.4960.100+24.11.oss).
   + I don't know.

**Comment by yukawa** [comment]
Thank you for the report. I've moved this to discussion to make it clear that you are looking for help from some other experts in this area.

While I'm also not familiar with IBus and GTK at all, if I had to look into this I would probably start from checking differences between `ibus-anthy` and `ibus-mozc`, especially about the differences on how IBus APIs are used.

**Comment by yukawa** [comment]
I'm not sure if it shares the same root cause or not, but the following thread looks quite similar to your case. just fyi.
 * https://github.com/ibus/ibus/issues/2637

**Reply by pineapplehunter** [reply]
This may be the case.
I tried downgrading ibus, and the problem was fixed.
ibus version with the problem: 1.5.29
ibus version without problem: 1.5.28

I will try to debug the issue.

**Reply by pineapplehunter** [reply, marked as answer]
I figured it out!
The issue was fixed in ibus 1.5.30, so upgrading ibus fixed the issue.

The issue was introduced on https://github.com/ibus/ibus/commit/e0595365c717cbc87ecb13bbf143fca2a64d1af1 which is between commits 1.5.28..1.5.29 and fixed on https://github.com/ibus/ibus/commit/719792d300579c1bfdf43251a83c6ed4e5594c07 which is between commits 1.5.29..1.5.30.
